File: reversi_cs/reversi.cpp
// #include <math.h>
// #include <stddef.h>
// #include <stdio.h>
// #include <unistd.h>

// #include <mutex>
// #include <cstdint>
#include <math.h>
#include <unistd.h>

#include <mutex>
#include <thread>
#include <vector>

// typedef unsigned long long uint64_t;
// typedef uint64_t ulong;
typedef unsigned long long ulong;
// typedef unsigned long long uint32_t;

// x, y から通し番号を得る
int coordinateToIndex(int x, int y) { return y * 8 + x; }

const ulong board_mask_4x6 =
    1ull << coordinateToIndex(1, 2) | 1ull << coordinateToIndex(1, 3) |
    1ull << coordinateToIndex(1, 4) | 1ull << coordinateToIndex(1, 5) |
    1ull << coordinateToIndex(2, 2) | 1ull << coordinateToIndex(2, 3) |
    1ull << coordinateToIndex(2, 4) | 1ull << coordinateToIndex(2, 5) |
    1ull << coordinateToIndex(3, 2) | 1ull << coordinateToIndex(3, 3) |
    1ull << coordinateToIndex(3, 4) | 1ull << coordinateToIndex(3, 5) |
    1ull << coordinateToIndex(4, 2) | 1ull << coordinateToIndex(4, 3) |
    1ull << coordinateToIndex(4, 4) | 1ull << coordinateToIndex(4, 5) |
    1ull << coordinateToIndex(5, 2) | 1ull << coordinateToIndex(5, 3) |
    1ull << coordinateToIndex(5, 4) | 1ull << coordinateToIndex(5, 5) |
    1ull << coordinateToIndex(6, 2) | 1ull << coordinateToIndex(6, 3) |
    1ull << coordinateToIndex(6, 4) | 1ull << coordinateToIndex(6, 5);

const ulong board_mask_4x5 =
    1ull << coordinateToIndex(1, 2) | 1ull << coordinateToIndex(1, 3) |
    1ull << coordinateToIndex(1, 4) | 1ull << coordinateToIndex(1, 5) |
    1ull << coordinateToIndex(2, 2) | 1ull << coordinateToIndex(2, 3) |
    1ull << coordinateToIndex(2, 4) | 1ull << coordinateToIndex(2, 5) |
    1ull << coordinateToIndex(3, 2) | 1ull << coordinateToIndex(3, 3) |
    1ull << coordinateToIndex(3, 4) | 1ull << coordinateToIndex(3, 5) |
    1ull << coordinateToIndex(4, 2) | 1ull << coordinateToIndex(4, 3) |
    1ull << coordinateToIndex(4, 4) | 1ull << coordinateToIndex(4, 5) |
    1ull << coordinateToIndex(5, 2) | 1ull << coordinateToIndex(5, 3) |
    1ull << coordinateToIndex(5, 4) | 1ull << coordinateToIndex(5, 5);

const ulong board_mask_4x4 =
    1ull << coordinateToIndex(2, 2) | 1ull << coordinateToIndex(2, 3) |
    1ull << coordinateToIndex(2, 4) | 1ull << coordinateToIndex(2, 5) |
    1ull << coordinateToIndex(3, 2) | 1ull << coordinateToIndex(3, 3) |
    1ull << coordinateToIndex(3, 4) | 1ull << coordinateToIndex(3, 5) |
    1ull << coordinateToIndex(4, 2) | 1ull << coordinateToIndex(4, 3) |
    1ull << coordinateToIndex(4, 4) | 1ull << coordinateToIndex(4, 5) |
    1ull << coordinateToIndex(5, 2) | 1ull << coordinateToIndex(5, 3) |
    1ull << coordinateToIndex(5, 4) | 1ull << coordinateToIndex(5, 5);

// 行と列（本当は8x8だが6x6でテスト中）
const int rows = 8;
const int columns = 8;

const int yNum = 8;
const int xNum = 8;

// 8方向の検索用ベクトル
const int searchVec[8][2] = {{-1, -1}, {+0, -1}, {+1, -1}, {-1, +0},
                             {+1, +0}, {-1, +1}, {+0, +1}, {+1, +1}};

enum ePLAYER { ePLAYER_P0 = 0, ePLAYER_P1, NUM };
ePLAYER player = ePLAYER_P0;

ulong piece[] = {0, 0};
// ulong placeable[] = {0, 0};
ulong board_mask = 0;

int coordinateToIndex(int x, int y);
void thread_test0(uint32_t, bool);
void multi_thread(void (*func)(uint32_t, bool), uint32_t loop_count,
                  uint32_t concurrency, bool lock = true);
int simuration(ulong piece_player, ulong piece_rybal);

bool isPlaceable(int x, int y, ulong piece_player, ulong piece_rybal);
void Place(int x, int y, ulong &piece_player, ulong &piece_rybal);

struct MyTimer {
    clock_t start;
    MyTimer() { start = clock(); }
    virtual ~MyTimer()
    {
        clock_t end = clock();
        printf("経過時間 = %fsec.\n", (double)(end - start) / CLOCKS_PER_SEC);
    }
};

int main(void)
{
    // CPUの並列度（△コア，〇スレッドの〇）
    uint32_t concurrency = std::thread::hardware_concurrency();

    // 先手
    player = ePLAYER_P0;

#if false
    ulong disable_mask = 0;

    // 先手（白）
    piece[0] = 1ull << coordinateToIndex(0, 0) |
               1ull << coordinateToIndex(3, 3) |
               1ull << coordinateToIndex(3, 4);
    // 後手（黒）
    piece[1] =
        1ull << coordinateToIndex(0, 1) | 1ull << coordinateToIndex(0, 2) |
        1ull << coordinateToIndex(0, 3) | 1ull << coordinateToIndex(0, 4) |
        1ull << coordinateToIndex(0, 5) | 1ull << coordinateToIndex(1, 0) |
        1ull << coordinateToIndex(1, 1) | 1ull << coordinateToIndex(1, 2) |
        1ull << coordinateToIndex(1, 3) | 1ull << coordinateToIndex(1, 4) |
        1ull << coordinateToIndex(1, 5) | 1ull << coordinateToIndex(2, 1) |
        1ull << coordinateToIndex(2, 2) | 1ull << coordinateToIndex(2, 3) |
        // 1ull << coordinateToIndex(2, 4) | 1ull << coordinateToIndex(2, 5) |
        1ull << coordinateToIndex(3, 2);
#else
    // 先手（黒）
    piece[0] = 1ull << coordinateToIndex(rows / 2 - 1, columns / 2 - 1) |
               1ull << coordinateToIndex(rows / 2 - 0, columns / 2 - 0);
    // 後手（白）
    piece[1] = 1ull << coordinateToIndex(rows / 2 - 0, columns / 2 - 1) |
               1ull << coordinateToIndex(rows / 2 - 1, columns / 2 - 0);
#endif

    for (int y = 0; y < columns; y++) {
        for (int x = 0; x < rows; x++) {
            int put_bit = coordinateToIndex(x, y);
            uint64_t put_mask = 1ull << put_bit;
            board_mask |= put_mask;
        }
    }

    // multi_thread(thread_test0, 0, concurrency);

    {
        MyTimer myTimer;
        int variation = simuration(piece[0], piece[1]);
    }

    return 0;
}

void thread_test0(uint32_t, bool)
{
    while (1) {
    }

    //  sleep(1);
}

void multi_thread(void (*func)(uint32_t, bool), uint32_t loop_count,
                  uint32_t concurrency, bool lock)
{
    // MyTimer myTimer;

    std::vector<std::thread> threads;

    for (int i = 0; i < concurrency; i++) {
        threads.emplace_back(std::thread(func, loop_count, lock));
    }

    for (auto &&thread : threads) {
        thread.join();
    }
}

int simuration(ulong piece_player, ulong piece_rybal)
{
    int num = 0;

    for (int y = 0; y < yNum; y++) {
        for (int x = 0; x < xNum; x++) {
            if (isPlaceable(x, y, piece_player, piece_rybal)) {
                // int piece_bit = coordinateToIndex(x, y);
                // ulong piece_mask = 1ull << piece_bit;
                ulong temp_player = piece_player;
                ulong temp_rybal = piece_rybal;
                Place(x, y, temp_player, temp_rybal);
                num += simuration(temp_rybal, temp_player);
            }
        }
    }

    if (num == 0) {
        ulong tmp = piece_player;
        piece_player = piece_rybal;
        piece_rybal = tmp;

        for (int y = 0; y < yNum; y++) {
            for (int x = 0; x < xNum; x++) {
                if (isPlaceable(x, y, piece_player, piece_rybal)) {
                    // int piece_bit = coordinateToIndex(x, y);
                    // ulong piece_mask = 1ull << piece_bit;
                    ulong temp_player = piece_player;
                    ulong temp_rybal = piece_rybal;
                    Place(x, y, temp_player, temp_rybal);
                    num += simuration(temp_rybal, temp_player);
                }
            }
        }
    }

    return num == 0 ? 1 : num;
    // 置けるセルの数を返す
    // return GetBitCount(placeable[(int)player]);
}

bool bitTest(int x, int y, ulong mask)
{
    int piece_bit = coordinateToIndex(x, y);
    ulong piece_mask = 1ull << piece_bit;
    return (mask & piece_mask) == piece_mask;
}

bool IsExist(int x, int y, ulong mask) { return bitTest(x, y, mask); }

bool isPlaceable(int x, int y, ulong piece_player, ulong piece_rybal)
{
    // ePLAYER rybal = player == ePLAYER.P0 ? ePLAYER.P1 : ePLAYER.P0;

    // playerまたはrybalがすでに置いているセル
    if (bitTest(x, y, piece_player) || bitTest(x, y, piece_rybal) ||
        bitTest(x, y, ~board_mask)) {
        return false;
    }

    // 8方向にそれぞれに繰り返す
    // foreach (var v in searchVec) {
    for (int i = 0; i < 8; i++) {
        int step = 0;  // 検索の状態
                       // （0:挟まれる相手の駒を探索中，1:挟む自分の駒を探索中）

        int v_X = searchVec[i][0];
        int v_Y = searchVec[i][1];

        for (int dx = x + v_X, dy = y + v_Y;
             0 <= dx && dx < xNum && 0 <= dy && dy < yNum;
             dx += v_X, dy += v_Y) {
            if (step == 0) {
                if (bitTest(dx, dy, piece_rybal)) {
                    step++;
                    continue;
                } else {
                    break;
                }
            } else if (step == 1) {
                if (bitTest(dx, dy, piece_player)) {
                    return true;
                } else if (bitTest(dx, dy, piece_rybal)) {
                    continue;
                } else {
                    break;
                }
            }
        }
    }

    return false;
}

void Place(int x, int y, ulong &piece_player, ulong &piece_rybal)
{
    int piece_bit = coordinateToIndex(x, y);
    ulong piece_mask = 1ull << piece_bit;

    piece_player |= piece_mask;

    // foreach (var v in searchVec) {
    for (int i = 0; i < 8; i++) {
        int step = 0;

        int v_X = searchVec[i][0];
        int v_Y = searchVec[i][1];

        for (int dx = x + v_X, dy = y + v_Y;
             0 <= dx && dx < xNum && 0 <= dy && dy < yNum;
             dx += v_X, dy += v_Y) {
            if (step == 0) {
                if (IsExist(dx, dy, piece_rybal)) {
                    step++;
                    continue;
                } else {
                    break;
                }
            } else if (step == 1) {
                if (IsExist(dx, dy, piece_player)) {
                    while (0 <= dx && dx < xNum && 0 <= dy && dy < yNum &&
                           !(dx == x && dy == y)) {
                        piece_bit = coordinateToIndex(dx, dy);
                        piece_mask = 1ul << piece_bit;
                        piece_player |= piece_mask;
                        piece_rybal = (piece_rybal & ~piece_mask);

                        dx -= v_X;
                        dy -= v_Y;
                    }
                    break;
                } else if (IsExist(dx, dy, piece_rybal)) {
                    continue;
                } else {
                    break;
                }
            }
        }
    }
}

File: reversi_cs/Reversi.cs
using System.Numerics;

namespace reversi
{
    internal class Reversi
    {
        // 行列(x,y)
        public static int rows;
        public static int columns;
        static int simulation_hierarchy_max = 4;

        // 手番
        public enum ePLAYER { P0 = 0, P1, NUM };
        ePLAYER player = ePLAYER.P0;
        public ePLAYER Player { get { return player; } }

        // ePLAYER.P0とePLAYER.P1が置いたセルを表すbitmask
        ulong[] board = new ulong[(int)ePLAYER.NUM];

        // ePLAYER.P0とePLAYER.P1が置けるセルを表すbitmask
        ulong[] placeable = new ulong[(int)ePLAYER.NUM];

        // 両者置けるセルがなくなったときtrue
        bool isGameOver = false;

        public CResult[,] result = new CResult[8, 8];


        public Reversi(int row, int column)
        {
            // 盤面の数
            rows = row;
            columns = column;

            for (int y = 0; y < columns; y++) {
                for (int x = 0; x < rows; x++) {
                    int put_bit = coordinateToIndex(x, y);
                    ulong put_mask = 1ul << put_bit;
                    empty_board |= put_mask;
                }
            }

            // 先手
            player = ePLAYER.P0;

            // 先手（黒）
            board[0] = 1ul << coordinateToIndex(rows / 2 - 0, columns / 2 - 1) |
                       1ul << coordinateToIndex(rows / 2 - 1, columns / 2 - 0);
            // 後手（白）
            board[1] = 1ul << coordinateToIndex(rows / 2 - 1, columns / 2 - 1) |
                       1ul << coordinateToIndex(rows / 2 - 0, columns / 2 - 0);

            // 置ける位置を調べる
            PlaceableTest();

            isGameOver = false;
        }

        int GetNumberOfTrailingZeros(ulong x)
        {
            return BitOperations.TrailingZeroCount(x);
        }

        CResult simulationSingle(ulong[] board, int player, int hierarchy_max)
        {
            CResult result = new CResult();

            if (hierarchy_max <= 0) {
                result.set(board);
                return result;
            }

            ulong legalBoard = makeLegalBoard(board, player);

            if (legalBoard != 0ul) {
                int opponent = player ^ 1;
                ulong m = legalBoard;
                int bit;
                while ((bit = GetNumberOfTrailingZeros(m)) != 64) {
                    ulong[] temp_board = { board[0], board[1] };
                    reverse(1ul << bit, temp_board, player);
                    result.marge(simulationSingle(temp_board, opponent, hierarchy_max - 1));

                    m &= ~(1ul << bit);
                }
            } else {
                player ^= 1;

                legalBoard = makeLegalBoard(board, player);

                if (legalBoard != 0ul) {
                    int opponent = player ^ 1;
                    ulong m = legalBoard;
                    int bit;
                    while ((bit = GetNumberOfTrailingZeros(m)) != 64) {
                        ulong[] temp_board = { board[0], board[1] };
                        reverse(1ul << bit, temp_board, player);
                        result.marge(simulationSingle(temp_board, opponent, hierarchy_max - 1));

                        m &= ~(1ul << bit);
                    }
                } else {
                    result.set(board);
                }
            }

            return result;
        }

        private static bool bitTest(int x, int y, ulong mask)
        {
            int piece_bit = coordinateToIndex(x, y);
            ulong piece_mask = 1ul << piece_bit;
            return (mask & piece_mask) == piece_mask;
        }

        // playerが置いているセルか調べる
        public bool IsExist(int x, int y, ePLAYER player)
        {
            return bitTest(x, y, board[(int)player]);
        }

        // playerが置けるセルか調べる
        public bool IsPlaceable(int x, int y, ePLAYER player)
        {
            return bitTest(x, y, placeable[(int)player]);
        }

        // 置ける位置を調べる
        public int PlaceableTest()
        {
            placeable[0] = placeable[1] = 0;

            placeable[(int)player] = makeLegalBoard(board, (int)player);

            for (int x = 0; x < rows; x++) {
                for (int y = 0; y < rows; y++) {
                    int opponent = (int)player ^ 1;
                    ulong[] temp_board = { board[0], board[1] };
                    int bit = coordinateToIndex(x, y);
                    reverse(1ul << bit, temp_board, (int)player);
                    result[x, y] = simulationSingle(temp_board, opponent, simulation_hierarchy_max);
                }
            }

            // 置けるセルの数を返す
            return GetBitCount(placeable[(int)player]);
        }

        // xとyで指定したセルに置けるか調べる
        public bool IsPlaceable(int x, int y)
        {
            ulong legalBoard = makeLegalBoard(board, (int)player);
            return bitTest(x, y, legalBoard);
        }

        // xとyで指定したセルに置く
        public void Place(int x, int y)
        {
            int put_bit = coordinateToIndex(x, y);
            ulong put_mask = 1ul << put_bit;
            reverse(put_mask, board, (int)player);

            player = player == ePLAYER.P0 ? ePLAYER.P1 : ePLAYER.P0;

            if (0 == PlaceableTest()) {
                player = player == ePLAYER.P0 ? ePLAYER.P1 : ePLAYER.P0;

                if (0 == PlaceableTest()) {
                    isGameOver = true;
                }
            }
        }

        public int Places(ePLAYER player)
        {
            return GetBitCount(board[(int)player]);
        }

        public static int GetBitCount(ulong x)
        {
            return BitOperations.PopCount(x);
        }

        // x, y から通し番号を得る
        private static int coordinateToIndex(int x, int y)
        {
            //return y * rows + x;
            return y * 8 + x;
        }

        void reverse(ulong put_mask, ulong[] board, int player)
        {
            int opponent = player ^ 1;

            // 着手した場合のボードを生成
            ulong rev = 0;
            for (int i = 0; i < 8; i++) {
                ulong rev_ = 0;
                ulong mask = transfer(put_mask, i);
                while ((mask != 0) && ((mask & board[opponent]) != 0)) {
                    rev_ |= mask;
                    mask = transfer(mask, i);
                }
                if ((mask & board[player]) != 0) {
                    rev |= rev_;
                }
            }

            // 反転する
            board[player] ^= put_mask | rev;
            board[opponent] ^= rev;
        }

        public static ulong transfer(ulong put, int k)
        {
            switch (k) {
                case 0:  // 上
                    return (put << 8) & 0xffffffffffffff00;
                case 1:  // 右上
                    return (put << 7) & 0x7f7f7f7f7f7f7f00;
                case 2:  // 右
                    return (put >> 1) & 0x7f7f7f7f7f7f7f7f;
                case 3:  // 右下
                    return (put >> 9) & 0x007f7f7f7f7f7f7f;
                case 4:  // 下
                    return (put >> 8) & 0x00ffffffffffffff;
                case 5:  // 左下
                    return (put >> 7) & 0x00fefefefefefefe;
                case 6:  // 左
                    return (put << 1) & 0xfefefefefefefefe;
                case 7:  // 左上
                    return (put << 9) & 0xfefefefefefefe00;
                default:
                    return 0;
            }
        }

        ulong empty_board = 0;

        ulong makeLegalBoard(ulong[] board, int player)
        {
            int opponent = player ^ 1;

            // 左右端の番人
            ulong horizontalWatchBoard = board[opponent] & 0x7e7e7e7e7e7e7e7e;
            // 上下端の番人
            ulong verticalWatchBoard = board[opponent] & 0x00FFFFFFFFFFFF00;
            // 全辺の番人
            ulong allSideWatchBoard = board[opponent] & 0x007e7e7e7e7e7e00;
            // 空きマスのみにビットが立っているボード
            ulong blankBoard = empty_board & ~(board[player] | board[opponent]);
            // 隣に相手の色があるかを一時保存する
            ulong tmp;
            // 返り値
            ulong legalBoard;

            // 8方向チェック
            //  ・一度に返せる石は最大6つ
            //  ・高速化のためにforを展開(ほぼ意味ないけどw)
            // 左
            tmp = horizontalWatchBoard & (board[player] << 1);
            tmp |= horizontalWatchBoard & (tmp << 1);
            tmp |= horizontalWatchBoard & (tmp << 1);
            tmp |= horizontalWatchBoard & (tmp << 1);
            tmp |= horizontalWatchBoard & (tmp << 1);
            tmp |= horizontalWatchBoard & (tmp << 1);
            legalBoard = blankBoard & (tmp << 1);

            // 右
            tmp = horizontalWatchBoard & (board[player] >> 1);
            tmp |= horizontalWatchBoard & (tmp >> 1);
            tmp |= horizontalWatchBoard & (tmp >> 1);
            tmp |= horizontalWatchBoard & (tmp >> 1);
            tmp |= horizontalWatchBoard & (tmp >> 1);
            tmp |= horizontalWatchBoard & (tmp >> 1);
            legalBoard |= blankBoard & (tmp >> 1);

            // 上
            tmp = verticalWatchBoard & (board[player] << 8);
            tmp |= verticalWatchBoard & (tmp << 8);
            tmp |= verticalWatchBoard & (tmp << 8);
            tmp |= verticalWatchBoard & (tmp << 8);
            tmp |= verticalWatchBoard & (tmp << 8);
            tmp |= verticalWatchBoard & (tmp << 8);
            legalBoard |= blankBoard & (tmp << 8);

            // 下
            tmp = verticalWatchBoard & (board[player] >> 8);
            tmp |= verticalWatchBoard & (tmp >> 8);
            tmp |= verticalWatchBoard & (tmp >> 8);
            tmp |= verticalWatchBoard & (tmp >> 8);
            tmp |= verticalWatchBoard & (tmp >> 8);
            tmp |= verticalWatchBoard & (tmp >> 8);
            legalBoard |= blankBoard & (tmp >> 8);

            // 右斜め上
            tmp = allSideWatchBoard & (board[player] << 7);
            tmp |= allSideWatchBoard & (tmp << 7);
            tmp |= allSideWatchBoard & (tmp << 7);
            tmp |= allSideWatchBoard & (tmp << 7);
            tmp |= allSideWatchBoard & (tmp << 7);
            tmp |= allSideWatchBoard & (tmp << 7);
            legalBoard |= blankBoard & (tmp << 7);

            // 左斜め上
            tmp = allSideWatchBoard & (board[player] << 9);
            tmp |= allSideWatchBoard & (tmp << 9);
            tmp |= allSideWatchBoard & (tmp << 9);
            tmp |= allSideWatchBoard & (tmp << 9);
            tmp |= allSideWatchBoard & (tmp << 9);
            tmp |= allSideWatchBoard & (tmp << 9);
            legalBoard |= blankBoard & (tmp << 9);

            // 右斜め下
            tmp = allSideWatchBoard & (board[player] >> 9);
            tmp |= allSideWatchBoard & (tmp >> 9);
            tmp |= allSideWatchBoard & (tmp >> 9);
            tmp |= allSideWatchBoard & (tmp >> 9);
            tmp |= allSideWatchBoard & (tmp >> 9);
            tmp |= allSideWatchBoard & (tmp >> 9);
            legalBoard |= blankBoard & (tmp >> 9);

            // 左斜め下
            tmp = allSideWatchBoard & (board[player] >> 7);
            tmp |= allSideWatchBoard & (tmp >> 7);
            tmp |= allSideWatchBoard & (tmp >> 7);
            tmp |= allSideWatchBoard & (tmp >> 7);
            tmp |= allSideWatchBoard & (tmp >> 7);
            tmp |= allSideWatchBoard & (tmp >> 7);
            legalBoard |= blankBoard & (tmp >> 7);

            return legalBoard;
        }
    }
}
